File: 백준/Silver/11055. 가장 큰 증가 부분 수열/가장 큰 증가 부분 수열.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var resultCombination = [][]int{}

func nCr(n int, nums, ans []int, r int) {
	if n == len(nums) {
		if len(ans) == r {
			var tmp = []int{}
			for _, v := range ans {
				tmp = append(tmp, v)
			}

			resultCombination = append(resultCombination, tmp)
		}
		return
	}
	ans = append(ans, nums[n])
	nCr(n+1, nums, ans, r)
	ans = ans[:len(ans)-1]
	nCr(n+1, nums, ans, r)
}

func main() {

	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n int
	fmt.Fscanf(reader, "%d\n", &n)
	var numbers = make([]int, n)
	var res = make([]int, n)
	var max int
	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%d ", &numbers[i])
	}
	for i := 0; i < n; i++ {
		for j := 0; j < i; j++ {
			if numbers[i] > numbers[j] && res[i] < res[j] {
				res[i] = res[j]
			}
		}
		res[i] += numbers[i]

	}
	for _, v := range res {
		if max < v {
			max = v
		}
	}
	fmt.Fprintf(writer, "%d", max)
}

File: 백준/Silver/11055. 가장 큰 증가 부분 수열/가장 큰 증가 부분 수열.py
import re
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
n = int(input())
numbers = list(map(int, input().split()))
res=[0]*(n)
#가장 크고 싶다면 : 합을 계속 비교하면서 늘려야함
#만약 작고 작다면 -> 제낌 만약 크고 크다면 -> 넣음
#ex. 1 2 3 4 5 6 7 8 9 55 10 11 12 13 14 15 16 17 17 18 20
#55를 빼고 나머지를 넣는 것이 더 큼
#각 idx까지 증가하는 수열을 만들 경우의 합 저장
tmp_sum = 0
last_idx = 0
for i in range (n):
    for j in range (i):
        if numbers[i]>numbers[j] and res[i] < res[j] :
            res[i] = res[j]
    res[i] += numbers[i]
print(max(res))